opengovsg/plumber
# PLU-125: [EXCEL-9] Pre-request checks

## Problem
Before we make a request to M365, we need to ensure that:
1. We obey rate limits
2. We reflect usage in a DD dashboard + alerts for GIG

## Solution
This PR implements the 3 preflight checks:

* **Rate Limiting:** We will integrate rate-limiter-flexible with `beforeRequest`.
* **Usage monitoring:** We will log to DD and use DD to generate a dashboard

Note that some failures can be retried (specifically: rate limit and redis lock failures). I've re-factored `RetriableError` for this purpose, and allowed action / trigger  code to throw this error as well (previously it was limited to our workers).

## Tests
- Set rate limit to 1, and check that I am rate limited + a `RetriableError` is thrown
- Set multiple rate limits to 1, and check that the rate limit error with the _longest_ cooldown is thrown.
- Check that logging calls are made whenever I make a M365 call.
- Regression test: check that I still get `HttpError` on normal HTTP errors.
